### PR TITLE
Ampache XML API and minor tweaks

### PR DESCRIPTION
**Pull request by Faldon**
Ampache XML API
===========
- Implemented the missing album action
- Implemented the missing add and update parameters
- Implemented the missing limit and offset parameters

Database schema
==============
- Added disc number field to the albums table

Frontend
=======
- Try to extract album art from first track of the album if no cover file is given
- Loading cover art in AngularJS albumart directive
- Displaying albums under album artist, if possible

General changes
=============
- Increased the patch version number to handle the database schema update
- Updated scanner to look for more metadata when scanning the files
- Detecting album artist information

### CODE REVIEW

File: appinfo/update.php [hunk 1]
```diff
@@ -43,3 +43,15 @@
...
 	$query->execute();
 }
+
+if (version_compare($installedVersion, '0.3.12', '<')) {
+	$sqls = array(
+		'DELETE FROM `*PREFIX*music_artists`;',
+		'DELETE FROM `*PREFIX*music_albums`;',
+		'DELETE FROM `*PREFIX*music_tracks`;',
+		'DELETE FROM `*PREFIX*music_ampache_sessions`',
+		'DROP TABLE `*PREFIX*music_album_artists`;',
```
Comment (MorrisJobke) [inline appinfo/update.php]: Why do you drop this table?

File: controller/ampachecontroller.php [hunk 2]
```diff
@@ -224,12 +226,18 @@ protected function artists() {
...
 		$fuzzy = !((boolean) $this->params('exact'));
 
-		// TODO add & update
+		// range
+		$limit = $this->params('limit');
+		$offset = $this->params('offset');
+
+		// time spans
+        $add = $this->params('add');
+        $update = $this->params('update');
```
Comment (MorrisJobke) [inline controller/ampachecontroller.php]: The indentation is off. Please use tabs to indent :)

File: controller/ampachecontroller.php [hunk 3]
```diff
@@ -250,9 +258,13 @@ protected function artist_albums() {
 		$userId = $this->ampacheUser->getUserId();
 		$artistId = $this->params('filter');
 
+        // range
+        $limit = $this->params('limit');
+        $offset = $this->params('offset');
```
Comment (MorrisJobke) [inline controller/ampachecontroller.php]: Same here.

File: controller/ampachecontroller.php [hunk 4]
```diff
@@ -341,21 +353,18 @@ protected function songs() {
 		$filter = $this->params('filter');
 		$fuzzy = !((boolean) $this->params('exact'));
 
-		// TODO add & update
+        // range
+        $limit = $this->params('limit');
+        $offset = $this->params('offset');
```
Comment (MorrisJobke) [inline controller/ampachecontroller.php]: Same here.

File: controller/ampachecontroller.php [hunk 5]
```diff
@@ -394,19 +403,70 @@ protected function search_songs() {
...
 		$userId = $this->ampacheUser->getUserId();
 
 		// filter
 		$filter = $this->params('filter');
 		$fuzzy = !((boolean) $this->params('exact'));
 
-		// TODO add & update
+        // range
+        $limit = $this->params('limit');
+        $offset = $this->params('offset');
```
Comment (MorrisJobke) [inline controller/ampachecontroller.php]: Same here.

File: controller/apicontroller.php [hunk 6]
```diff
@@ -107,6 +107,11 @@ public function collection() {
 		$artists = array();
 		foreach ($allTracks as $track) {
 			$artist = &$allArtistsById[$track->getArtistId()];
+            $albumArtists = $this->albumBusinessLayer->find($track->getAlbumId(), $this->userId);
```
Comment (MorrisJobke) [inline controller/apicontroller.php]: Indentation.

**Comment by MorrisJobke** [comment]
Hey @Faldon :) Thanks for your contribution :)

This are a lot of changes. Would it be possible to do them as separate PRs? THis makes the reviewing easier and changes that are good to merge could get in easier. Otherwise I would cherry pick some commits already to get them in and let you fix the other stuff.

What way to you want to go? :)

**Comment by MorrisJobke** [comment]
The changes in general look good :)

Also check the failing CI jobs ;)

**Comment by Faldon** [comment]
Hey @MorrisJobke 
Sorry for the indentation mess, I had settings for a different project active.
I would split the ampache api changes, the frontend changes and the database & general changes into separate pull requests. 

I'm not familiar with Travis CI, but I'll have a look into it.

**Pull request closed without merging**